File: src/main/scala/com/github/mauricio/postgresql/column/TimestampEncoderDecoder.scala
package com.github.mauricio.postgresql.column

import org.joda.time.format.DateTimeFormat
import org.joda.time.{ReadableDateTime, DateTime}
import com.github.mauricio.postgresql.exceptions.DateEncoderNotAvailableException
import java.util.{Calendar, Date}
import java.sql.Timestamp

/**
 * User: Maurício Linhares
 * Date: 3/5/12
 * Time: 6:10 PM
 */

object TimestampEncoderDecoder {
  val Instance = new TimestampEncoderDecoder()
}

class TimestampEncoderDecoder extends ColumnEncoderDecoder {

  private val format = DateTimeFormat.forPattern("yyyy-MM-dd HH:mm:ss.SSSSSS")

  def formatter = format

  override def decode(value: String): ReadableDateTime = {
    formatter.parseDateTime(value)
  }

  override def encode( value : Any ) : String = {
    value match {
      case t : Timestamp => this.formatter.print( new DateTime(t) )
      case t : Date => this.formatter.print( new DateTime(t) )
      case t : Calendar => this.formatter.print( new DateTime(t) )
      case t : ReadableDateTime => this.formatter.print(t)
      case _ => throw new DateEncoderNotAvailableException(value)
    }
  }

}

File: db-async-common/src/main/java/com/github/jasync/sql/db/column/TimestampEncoderDecoder.kt
package com.github.jasync.sql.db.column

import com.github.jasync.sql.db.exceptions.DateEncoderNotAvailableException
import java.sql.Timestamp
import java.time.LocalDateTime
import java.time.ZoneOffset
import java.time.format.DateTimeFormatterBuilder
import java.time.temporal.TemporalAccessor
import java.util.Calendar
import java.util.Date

open class TimestampEncoderDecoder : ColumnEncoderDecoder {
    companion object {
        const val BaseFormat = "yyyy-MM-dd HH:mm:ss"
        const val MillisFormat = ".SSSSSS"
        val Instance = TimestampEncoderDecoder()
    }

    private val optional = DateTimeFormatterBuilder()
        .appendPattern(MillisFormat).toFormatter()
    private val optionalTimeZone = DateTimeFormatterBuilder()
        .appendPattern("[X][Z]").toFormatter()

    private val builder = DateTimeFormatterBuilder()
        .appendPattern(BaseFormat)
        .appendOptional(optional)
        .appendOptional(optionalTimeZone)

    private val timezonedPrinter = DateTimeFormatterBuilder()
        .appendPattern("${BaseFormat}${MillisFormat}Z").toFormatter()

    private val nonTimezonedPrinter = DateTimeFormatterBuilder()
        .appendPattern("${BaseFormat}$MillisFormat").toFormatter()

    private val format = builder.toFormatter()

    // java.util.Dates are constructed using the system default timezone, replicate this behavior when encoding a legacy date
    private fun encodeLegacyDate(legacyDate: Date): String =
        legacyDate.toInstant().atOffset(ZoneOffset.UTC).format(this.timezonedPrinter)

    open fun formatter() = format

    override fun decode(value: String): Any {
        return LocalDateTime.parse(value, formatter())
    }

    override fun encode(value: Any): String {
        return when (value) {
            is Timestamp -> encodeLegacyDate(value)
            is Date -> encodeLegacyDate(value)
            is Calendar -> encodeLegacyDate(value.time)
            is LocalDateTime -> this.nonTimezonedPrinter.format(value)
            is TemporalAccessor -> this.timezonedPrinter.format(value)
            else -> throw DateEncoderNotAvailableException(value)
        }
    }
}
